ports/tcpport.js

```diff
@@ -240,7 +242,26 @@ class TcpPort extends EventEmitter {
         });
 
         // send buffer to slave
-        this._client.write(buffer);
+        let previousWritePromise = this._writeCompleted;
+        let newWritePromise = new Promise((resolveNewWrite, rejectNewWrite) => {
+            // Wait for the completion of any write that happened before.
+            previousWritePromise.finally(() => {
+                try {
+                    // The previous write succeeded, write the new buffer.
+                    if (this._client.write(buffer)) {
+                        // Mark this write as complete.
+                        resolveNewWrite();
+                    } else {
+                        // Wait for one `drain` event to mark this write as complete.
+                        this._client.once("drain", resolveNewWrite);
```

Comment:
do we need to address a case where "drain" is not called ? for example connection fail ?

Reply:
I don't think that's necessary. The [`drain` event documentation](https://nodejs.org/api/stream.html#event-drain) says

> If a call to [stream.write(chunk)](https://nodejs.org/api/stream.html#writablewritechunk-encoding-callback) returns false, the 'drain' event will be emitted when it is appropriate to resume writing data to the stream.

I think this means that we are guaranteed to get a `drain` event, if `Socket.write` returns `false`. You can also see that the usage example in the documentation does not include any specific consideration for the `error` event.

But in case a user of the `TcpPort` decides to close and then re-open the the port, I overwrote the promise in the `connect` event so that we are sure to not be blocked by writes on a closed port.